pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -1184,168 +898,127 @@ protected IdealState ensureAllPartitionsConsuming(final TableConfig tableConfig,
-      if (success) {
-        newSegmentNames.add(newLLCSegmentName.getSegmentName());
-      }
-    }
-    return newSegmentNames;
+    String rawTableName = TableNameBuilder.extractRawTableName(realtimeTableName);
+    LLCSegmentName newLLCSegmentName =
+        new LLCSegmentName(rawTableName, partitionId, STARTING_SEQUENCE_NUMBER, creationTimeMs);
+    String newSegmentName = newLLCSegmentName.getSegmentName();
+    long startOffset =
+        getPartitionOffset(partitionLevelStreamConfig, partitionLevelStreamConfig.getOffsetCriteria(), partitionId);
+    CommittingSegmentDescriptor committingSegmentDescriptor = new CommittingSegmentDescriptor(null, startOffset, 0);
+    createNewSegmentZKMetadata(tableConfig, partitionLevelStreamConfig, newLLCSegmentName, creationTimeMs,
+        committingSegmentDescriptor, null, instancePartitions, numPartitions, numReplicas);
+
+    return newSegmentName;
   }
 
   @VisibleForTesting
-  protected long getCurrentTimeMs() {
+  long getCurrentTimeMs() {
     return System.currentTimeMillis();
   }
 
-  protected IdealState updateIdealStateOnSegmentCompletion(@Nonnull IdealState idealState,
-      @Nonnull String currentSegmentId, @Nonnull String newSegmentId,
-      @Nonnull PartitionAssignment partitionAssignment) {
-
-    Map<String, List<String>> instanceAssignments;
-
-    RealtimeSegmentAssignmentStrategy strategy = new ConsumingSegmentAssignmentStrategy();
-    try {
-      instanceAssignments = strategy.assign(Lists.newArrayList(newSegmentId), partitionAssignment);
-    } catch (InvalidConfigException e) {
-      _controllerMetrics.addMeteredTableValue(idealState.getResourceName(),
-          ControllerMeter.CONTROLLER_REALTIME_TABLE_SEGMENT_ASSIGNMENT_ERROR, 1L);
-      throw new IllegalStateException("Caught exception when updating ideal state on segment completion", e);
+  private int getNumPartitionsFromIdealState(IdealState idealState) {
+    int numPartitions = 0;
+    for (String segmentName : idealState.getRecord().getMapFields().keySet()) {
+      if (LLCSegmentName.isLowLevelConsumerSegmentName(segmentName)) {
+        numPartitions = Math.max(numPartitions, new LLCSegmentName(segmentName).getPartitionId() + 1);
+      }
     }
+    return numPartitions;
+  }
 
-    List<String> newSegmentInstances = instanceAssignments.get(newSegmentId);
-    Set<String> currentSegmentInstances = idealState.getInstanceSet(currentSegmentId);
-    for (String instance : currentSegmentInstances) {
-      idealState.setPartitionState(currentSegmentId, instance,
-          PinotHelixSegmentOnlineOfflineStateModelGenerator.ONLINE_STATE);
+  private int getNumReplicas(TableConfig tableConfig, InstancePartitions instancePartitions) {
+    if (instancePartitions.getNumReplicaGroups() == 1) {
+      // Non-replica-group based
+      return tableConfig.getValidationConfig().getReplicasPerPartitionNumber();
+    } else {
+      // Replica-group based
+      return instancePartitions.getNumReplicaGroups();
```

Comment:
even if it is replica group based, shouldn't the numReplicas match the `tableConfig.getValidationConfig().getReplicasPerPartitionNumber()` ?

Reply:
Not always. If we assign instances with instanceAssignmentConfig, it can be different from  `tableConfig.getValidationConfig().getReplicasPerPartitionNumber()`. We will log warning when assigning instances if these 2 numbers not match.